apps/zotonic_mod_translation/src/support/translation_translate.erl

```diff
@@ -0,0 +1,154 @@
+%% can be trans records, a single string or a list of strings. If the source language
+%% is 'en' then the .po files are consulted for a preferred translation.
+%% Both the from and to language must be configured as editable languages.
+
+-spec translate(FromLanguage, ToLanguage, Texts, Context) -> {ok, Translations} | {error, Reason} when
+    FromLanguage :: z_language:language(),
+    ToLanguage :: z_language:language(),
+    Texts :: [ Text ] | Text,
+    Text :: binary() | #trans{},
+    Context :: z:context(),
+    Translations :: [ binary() | undefined ],
+    Reason :: term().
+translate(FromLanguage, ToLanguage, Texts, Context) ->
+    FromCode = to_lang(FromLanguage, Context),
+    ToCode = to_lang(ToLanguage, Context),
+    translate_1(FromCode, ToCode, to_list(Texts), Context).
+
+translate_1({error, _}, _ToCode, _Texts, _Context) ->
+    {error, from_language};
+translate_1(_FromCode, {error, _}, _Texts, _Context) ->
+    {error, to_language};
+translate_1({ok, FromCode}, {ok, ToCode}, Texts, Context) ->
+    PartlyTranslated = lists:map(
+        fun(Text) ->
+            local_trans(FromCode, ToCode, trim(Text), Context)
+        end,
+        Texts),
+    add_service_trans(FromCode, ToCode, PartlyTranslated, Context).
+
+
+add_service_trans(FromCode, ToCode, PartlyTranslated, Context) ->
+    case lists:filtermap(
+        fun
+            ({FromText, undefined}) -> {true, FromText};
+            (_) -> false
+        end,
+        PartlyTranslated)
+    of
+        [] ->
+            Translations = [ T || {_, T} <- PartlyTranslated ],
+            {ok, Translations};
+        ToTranslate ->
+            Notification = #translate{
+                from = FromCode,
+                to = ToCode,
+                texts = ToTranslate
+            },
+            case z_notifier:first(Notification, Context) of
+                {ok, Translations} ->
+                    {ok, merge_translations(PartlyTranslated, Translations, [])};
+                {error, _} = Error ->
+                    Error;
+                undefined ->
+                    {_, Translations} = lists:unzip(PartlyTranslated),
+                    {ok, Translations}
+            end
+    end.
+
+merge_translations([], [], Acc) ->
+    lists:reverse(Acc);
+merge_translations([{_Text, undefined}|Rest], [Tr|Trs], Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]);
+merge_translations([{_Text, Tr}|Rest], Trs, Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]).
+
+
+local_trans(FromCode, ToCode, Text, _Context) when is_binary(Text), FromCode =:= ToCode ->
+    {Text, Text};
+local_trans(_FromCode, _ToCode, <<>>, _Context) ->
+    {<<>>, <<>>};
+local_trans(_FromCode, _ToCode, #trans{ tr = [] }, _Context) ->
+    {<<>>, <<>>};
+local_trans(FromCode, ToCode, #trans{ tr = Tr }, Context) ->
+    FromCodeText = lists:keyfind(FromCode, 1, Tr),
+    case lists:keyfind(ToCode, 1, Tr) of
+        {_ToCode, ToText} ->
+            {<<>>, ToText};
```

Comment:
Shouldn't `<<>>` be `fromText`?

Reply:
The From here doesn't really matter anymore, as it is translated. You could argue that it could be the FromCodeText, but that could also be `false` as it might not be available in the `#trans` record.

We only need the FromText if the translation is `undefined` and we have to pass the text to a translation service.